File: java-code/src/test/java/com/vd/study/java_code/builder/BuilderTest.java
package com.vd.study.java_code.builder;

import org.junit.Assert;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collections;

public class BuilderTest {

    @Test
    public void employeeCreation_AfterEmployeeCreationUsingBuilderOnlyWithImportantFields_ReturnCorrectEmployeeObject() {
        Position position = new Position("Position name", 2.2);
        Education education = new Education("Place", "Profession", 1.1);
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPosition(position)
                .setExperience(5.4)
                .setEducation(new ArrayList<>(Collections.singletonList(education)));

        Employee employee = builder.build();

        Assert.assertEquals("Firstname", employee.firstname);
        Assert.assertEquals("Lastname", employee.lastname);
        Assert.assertNull(employee.patronymic);
        Assert.assertSame(position, employee.position);
        Assert.assertEquals(5.4, employee.experience, 0.0);
        Assert.assertSame(null, employee.address);
        Assert.assertEquals(1, employee.education.size());
        Assert.assertSame(education, employee.education.get(0));
    }

    @Test
    public void employeeCreation_AfterEmployeeCreationUsingBuilderAndSecondSetEducationFunctionInvoke_ReturnCorrectEducation() {
        Position position = new Position("Position name", 2.2);
        Address address = new Address("Area", "Street", "House number");
        Education education1 = new Education("Place1", "Profession1", 1.1);
        Education education2 = new Education("Place2", "Profession2", 2.2);
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setPosition(position)
                .setExperience(5.4)
                .setAddress(address)
                .setEducation(new ArrayList<>(Collections.singletonList(education1)))
                .setEducation(new ArrayList<>(Collections.singletonList(education2)));

        Employee employee = builder.build();

        Assert.assertEquals(1, employee.education.size());
        Assert.assertSame(education2, employee.education.get(0));
    }

    @Test
    public void employeeCreation_AfterEmployeeCreationUsingBuilderAndAddEducationFunction_ReturnCorrectEducation() {
        Position position = new Position("Position name", 2.2);
        Address address = new Address("Area", "Street", "House number");
        Education education1 = new Education("Place1", "Profession1", 1.1);
        Education education2 = new Education("Place2", "Profession2", 2.2);
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setPosition(position)
                .setExperience(5.4)
                .setAddress(address)
                .addEducation(education1)
                .addEducation(education2);

        Employee employee = builder.build();

        Assert.assertEquals(2, employee.education.size());
        Assert.assertSame(education1, employee.education.get(0));
        Assert.assertSame(education2, employee.education.get(1));
    }

    @Test
    public void employeeCreation_AfterEmployeeCreationUsingBuilder_ReturnCorrectEmployeeObject() {
        Position position = new Position("Position name", 2.2);
        Address address = new Address("Area", "Street", "House number");
        Education education = new Education("Place", "Profession", 1.1);
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setPosition(position)
                .setExperience(5.4)
                .setAddress(address)
                .setEducation(new ArrayList<>(Collections.singletonList(education)));

        Employee employee = builder.build();

        Assert.assertEquals("Firstname", employee.firstname);
        Assert.assertEquals("Lastname", employee.lastname);
        Assert.assertEquals("Patronymic", employee.patronymic);
        Assert.assertSame(position, employee.position);
        Assert.assertEquals(5.4, employee.experience, 0.0);
        Assert.assertSame(address, employee.address);
        Assert.assertEquals(1, employee.education.size());
        Assert.assertSame(education, employee.education.get(0));
    }

    @Test(expected = IllegalArgumentException.class)
    public void employeeCreation_AfterEmployeeCreationUsingBuilderWithoutFirstname_ThrowIllegalArgumentException() {
        Employee.Builder builder = new Employee.Builder()
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setPosition(new Position("Position name", 2.2))
                .setExperience(5.4)
                .setAddress(new Address("Area", "Street", "House number"))
                .setEducation(new ArrayList<>(Collections.singletonList(new Education("Place", "Profession", 1.1))));

        builder.build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void employeeCreation_AfterEmployeeCreationUsingBuilderWithoutLastname_ThrowIllegalArgumentException() {
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setPatronymic("Patronymic")
                .setPosition(new Position("Position name", 2.2))
                .setExperience(5.4)
                .setAddress(new Address("Area", "Street", "House number"))
                .setEducation(new ArrayList<>(Collections.singletonList(new Education("Place", "Profession", 1.1))));

        builder.build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void employeeCreation_AfterEmployeeCreationUsingBuilderWithoutPosition_ThrowIllegalArgumentException() {
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setExperience(5.4)
                .setAddress(new Address("Area", "Street", "House number"))
                .setEducation(new ArrayList<>(Collections.singletonList(new Education("Place", "Profession", 1.1))));

        builder.build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void employeeCreation_AfterEmployeeCreationUsingBuilderWithoutExperience_ThrowIllegalArgumentException() {
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setPosition(new Position("Position name", 2.2))
                .setAddress(new Address("Area", "Street", "House number"))
                .setEducation(new ArrayList<>(Collections.singletonList(new Education("Place", "Profession", 1.1))));

        builder.build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void employeeCreation_AfterEmployeeCreationUsingBuilderWithoutEducation_ThrowIllegalArgumentException() {
        Employee.Builder builder = new Employee.Builder()
                .setFirstname("Firstname")
                .setLastname("Lastname")
                .setPatronymic("Patronymic")
                .setPosition(new Position("Position name", 2.2))
                .setExperience(5.4)
                .setAddress(new Address("Area", "Street", "House number"));

        builder.build();
    }
}

File: kotlin-code/src/test/java/com/vd/study/kotlin_code/builder/BuilderTest.kt
package com.vd.study.kotlin_code.builder

import org.junit.Assert.assertEquals
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Test

class BuilderTest {

    @Test
    fun employeeCreation_AfterEmployeeCreationUsingBuilderOnlyWithImportantFields_ReturnCorrectEmployeeObject() {
        val position = Position("Position name", 2.2)
        val education = Education("Place", "Profession", 1.1)
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPosition(position)
            .setExperience(5.4)
            .setEducation(listOf(education))

        val employee = builder.build()

        assertEquals("Firstname", employee.firstname)
        assertEquals("Lastname", employee.lastname)
        assertEquals(null, employee.patronymic)
        assertSame(position, employee.position)
        assertTrue(5.4 == employee.experience)
        assertSame(null, employee.address)
        assertEquals(1, employee.education.size)
        assertSame(education, employee.education[0])
    }

    @Test
    fun employeeCreation_AfterEmployeeCreationUsingBuilderAndSecondSetEducationFunctionInvoke_ReturnCorrectEducation() {
        val position = Position("Position name", 2.2)
        val address = Address("Area", "Street", "House number")
        val education1 = Education("Place1", "Profession1", 1.1)
        val education2 = Education("Place2", "Profession2", 2.2)
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setPosition(position)
            .setExperience(5.4)
            .setAddress(address)
            .setEducation(listOf(education1))
            .setEducation(listOf(education2))

        val employee = builder.build()

        assertEquals(1, employee.education.size)
        assertSame(education2, employee.education[0])
    }

    @Test
    fun employeeCreation_AfterEmployeeCreationUsingBuilderAndAddEducationFunction_ReturnCorrectEducation() {
        val position = Position("Position name", 2.2)
        val address = Address("Area", "Street", "House number")
        val education1 = Education("Place1", "Profession1", 1.1)
        val education2 = Education("Place2", "Profession2", 2.2)
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setPosition(position)
            .setExperience(5.4)
            .setAddress(address)
            .addEducation(education1)
            .addEducation(education2)

        val employee = builder.build()

        assertEquals(2, employee.education.size)
        assertSame(education1, employee.education[0])
        assertSame(education2, employee.education[1])
    }

    @Test
    fun employeeCreation_AfterEmployeeCreationUsingBuilder_ReturnCorrectEmployeeObject() {
        val position = Position("Position name", 2.2)
        val address = Address("Area", "Street", "House number")
        val education = Education("Place", "Profession", 1.1)
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setPosition(position)
            .setExperience(5.4)
            .setAddress(address)
            .setEducation(listOf(education))

        val employee = builder.build()

        assertEquals("Firstname", employee.firstname)
        assertEquals("Lastname", employee.lastname)
        assertEquals("Patronymic", employee.patronymic)
        assertSame(position, employee.position)
        assertTrue(5.4 == employee.experience)
        assertSame(address, employee.address)
        assertEquals(1, employee.education.size)
        assertSame(education, employee.education[0])
    }

    @Test(expected = IllegalArgumentException::class)
    fun employeeCreation_AfterEmployeeCreationUsingBuilderWithoutFirstname_ThrowIllegalArgumentException() {
        val builder = Employee.Builder()
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setPosition(Position("Position name", 2.2))
            .setExperience(5.4)
            .setAddress(Address("Area", "Street", "House number"))
            .setEducation(listOf(Education("Place", "Profession", 1.1)))

        builder.build()
    }

    @Test(expected = IllegalArgumentException::class)
    fun employeeCreation_AfterEmployeeCreationUsingBuilderWithoutLastname_ThrowIllegalArgumentException() {
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setPatronymic("Patronymic")
            .setPosition(Position("Position name", 2.2))
            .setExperience(5.4)
            .setAddress(Address("Area", "Street", "House number"))
            .setEducation(listOf(Education("Place", "Profession", 1.1)))

        builder.build()
    }

    @Test(expected = IllegalArgumentException::class)
    fun employeeCreation_AfterEmployeeCreationUsingBuilderWithoutPosition_ThrowIllegalArgumentException() {
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setExperience(5.4)
            .setAddress(Address("Area", "Street", "House number"))
            .setEducation(listOf(Education("Place", "Profession", 1.1)))

        builder.build()
    }

    @Test(expected = IllegalArgumentException::class)
    fun employeeCreation_AfterEmployeeCreationUsingBuilderWithoutExperience_ThrowIllegalArgumentException() {
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setPosition(Position("Position name", 2.2))
            .setAddress(Address("Area", "Street", "House number"))
            .setEducation(listOf(Education("Place", "Profession", 1.1)))

        builder.build()
    }

    @Test(expected = IllegalArgumentException::class)
    fun employeeCreation_AfterEmployeeCreationUsingBuilderWithoutEducation_ThrowIllegalArgumentException() {
        val builder = Employee.Builder()
            .setFirstname("Firstname")
            .setLastname("Lastname")
            .setPatronymic("Patronymic")
            .setPosition(Position("Position name", 2.2))
            .setExperience(5.4)
            .setAddress(Address("Area", "Street", "House number"))

        builder.build()
    }
}